rust-runtime/aws-smithy-http-server-python/src/context/layer.rs

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! A [tower::Layer] for injecting and populating [PyContext].
+
+use std::task::{Context, Poll};
+
+use http::{Request, Response};
+use lambda_http::Context as LambdaContext;
+use tower::{Layer, Service};
+
+use crate::lambda::PyLambdaContext;
+
+use super::PyContext;
+
+/// AddPyContextLayer is a [tower::Layer] that populates given [PyContext] from the [Request]
+/// and injects [PyContext] to the [Request] as an extension.
+pub struct AddPyContextLayer {
+    ctx: PyContext,
+}
+
+impl AddPyContextLayer {
+    pub fn new(ctx: PyContext) -> Self {
+        Self { ctx }
+    }
+}
+
+impl<S> Layer<S> for AddPyContextLayer {
+    type Service = AddPyContextService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        AddPyContextService {
+            inner,
+            ctx: self.ctx.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct AddPyContextService<S> {
+    inner: S,
+    ctx: PyContext,
+}
+
+impl<ResBody, ReqBody, S> Service<Request<ReqBody>> for AddPyContextService<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        if self.ctx.has_lambda_context_fields() {
+            let py_lambda_ctx = req
+                .extensions()
+                .get::<LambdaContext>()
+                .cloned()
```

Comment:
Do we want to `.cloned()` this? What's the difference if we `.remove::<LambdaContext>()`?

Reply:
I didn't want to use `remove` because other layers might want to use `LambdaContext`, is it a good practice to remove extensions?